### PR TITLE
disable monotone constraint in objective functions with renew_tree_output

### LINKED ISSUE #3371: Monotone constraint with Quantile distribution does not work correctly
I tried quantile distribution while the monotone constraint is set and it looks it is not implemented correctly. The result prediction is not monotone, see the example and image bellow. 

## How you are using LightGBM?

* Python package

## Environment info

Operating System: Linux Debian 10.5 (x86-64)

Python version:  Python 3.7.3rc1 

LightGBM version or commit hash: lightgbm==2.3.1

## Reproducible example

```python
# prepare data
import numpy as np
np.random.seed(1)

def f(x):
    """The function to predict."""
    return x * np.sin(x)

X = np.atleast_2d(np.random.uniform(0, 10.0, size=100)).T
X = X.astype(np.float32)

y = f(X).ravel()

dy = 1.5 + 1.0 * np.random.random(y.shape)
noise = np.random.normal(0, dy)
y += noise
y = y.astype(np.float32)

xx = np.atleast_2d(np.linspace(0, 10, 100)).T
xx = xx.astype(np.float32)

# prepare plot function
import matplotlib.pyplot as plt

def plot_prediction_quantile(xx, fxx, xx_label, X, y, y_pred,ylim, title, y_upper=None, y_lower=None, 
                             confidence_label=None):
    fig = plt.figure()
    plt.plot(xx, fxx, 'g:', label=xx_label)
    if (X is not None) and (y is not None):
        plt.plot(X, y, 'b.', markersize=10, label=u'Observations')
    plt.plot(xx, y_pred, 'r-', label=u'Prediction')
    if (y_upper is not None) and (y_lower is not None):
        plt.plot(xx, y_upper, 'k-')
        plt.plot(xx, y_lower, 'k-')
        plt.fill(np.concatenate([xx, xx[::-1]]),
                 np.concatenate([y_upper, y_lower[::-1]]),
                 alpha=.5, fc='b', ec='None', label=confidence_label)
    plt.xlabel('$x$')
    plt.ylabel('$f(x)$')
    plt.ylim(ylim)
    plt.legend(loc='upper left')
    plt.title(title)
    plt.show()
    
# prepare lightgbm
from lightgbm import LGBMRegressor

lgb_params = {
    'n_jobs': 1,
    'max_depth': 5,
    'min_data_in_leaf': 3,
    'n_estimators': 100,
    'learning_rate': 0.1,
    'colsample_bytree': 0.9,
    'boosting_type': 'gbdt',
    'monotone_constraints': -1
}

lgb_no_monotonicity = LGBMRegressor(objective='quantile', alpha=0.4, **lgb_params)
lgb_no_monotonicity .fit(X, y)
y_no_monotonicity_lgb = lgb_no_monotonicity.predict(xx)

plot_prediction_quantile(xx, f(xx), r'f(x)$', X, y, y_no_monotonicity_lgb, [min(y)-0.2*min(y), max(y)+0.2*max(y)], "LightGBM Quantile (quantile_alpha=0.4) with monotone constraint 1 - Monotonicity constraint violated!")
```

## Result plot

lightgbm

## Steps to reproduce

1. run from console using

`python example.py`

2. see the resulting image

### PR DESCRIPTION
**Pull request by guolinke**
to fix #3371

### CODE REVIEW

File: src/boosting/gbdt.cpp [hunk 1]
```diff
@@ -71,6 +71,9 @@ void GBDT::Init(const Config* config, const Dataset* train_data, const Objective
   num_tree_per_iteration_ = num_class_;
   if (objective_function_ != nullptr) {
     num_tree_per_iteration_ = objective_function_->NumModelPerIteration();
+    if (objective_function_->IsRenewTreeOutput() && !config->monotone_constraints.empty()) {
+      Log::Fatal('Cannot use ``monotone_constraints`` in %s objective, please disable it.', objective_function_->GetName());
```
Comment (guolinke) [inline src/boosting/gbdt.cpp:75]:
```suggestion
      Log::Fatal("Cannot use ``monotone_constraints`` in %s objective, please disable it.", objective_function_->GetName());
```

File: src/boosting/gbdt.cpp [hunk 2]
```diff
@@ -657,6 +660,9 @@ void GBDT::ResetTrainingData(const Dataset* train_data, const ObjectiveFunction*
   objective_function_ = objective_function;
   if (objective_function_ != nullptr) {
     CHECK_EQ(num_tree_per_iteration_, objective_function_->NumModelPerIteration());
+    if (objective_function_->IsRenewTreeOutput() && !config_->monotone_constraints.empty()) {
+      Log::Fatal('Cannot use ``monotone_constraints`` in %s objective, please disable it.', objective_function_->GetName());
```
Comment (guolinke) [inline src/boosting/gbdt.cpp:664]:
```suggestion
      Log::Fatal("Cannot use ``monotone_constraints`` in %s objective, please disable it.", objective_function_->GetName());
```

File: src/boosting/gbdt.cpp [hunk 3]
```diff
@@ -710,6 +716,9 @@ void GBDT::ResetConfig(const Config* config) {
   if (!config->feature_contri.empty()) {
     CHECK_EQ(static_cast<size_t>(train_data_->num_total_features()), config->feature_contri.size());
   }
+  if (objective_function_ != nullptr && objective_function_->IsRenewTreeOutput() && !config->monotone_constraints.empty()) {
+    Log::Fatal('Cannot use ``monotone_constraints`` in %s objective, please disable it.', objective_function_->GetName());
```
Comment (guolinke) [inline src/boosting/gbdt.cpp:720]:
```suggestion
    Log::Fatal("Cannot use ``monotone_constraints`` in %s objective, please disable it.", objective_function_->GetName());
```

**Comment by StrikerRUS** [comment]
Hmm, I have never seen this error before:
```
================================== FAILURES ===================================
__________________ TestBasic.test_plot_split_value_histogram __________________

self = <test_plotting.TestBasic testMethod=test_plot_split_value_histogram>

    @unittest.skipIf(not MATPLOTLIB_INSTALLED, 'matplotlib is not installed')
    def test_plot_split_value_histogram(self):
>       gbm0 = lgb.train(self.params, self.train_data, num_boost_round=10)

..\tests\python_package_test\test_plotting.py:64: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Miniconda\envs\test-env\lib\site-packages\lightgbm\engine.py:231: in train
    booster = Booster(params=params, train_set=train_set)
C:\Miniconda\envs\test-env\lib\site-packages\lightgbm\basic.py:2010: in __init__
    self.__get_eval_info()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <lightgbm.basic.Booster object at 0x000000000A248B88>

    def __get_eval_info(self):
        """Get inner evaluation count and names."""
        if self.__need_reload_eval_info:
            self.__need_reload_eval_info = False
            out_num_eval = ctypes.c_int(0)
            # Get num of inner evals
            _safe_call(_LIB.LGBM_BoosterGetEvalCounts(
                self.handle,
                ctypes.byref(out_num_eval)))
            self.__num_inner_eval = out_num_eval.value
            if self.__num_inner_eval > 0:
                # Get name of evals
                tmp_out_len = ctypes.c_int(0)
                reserved_string_buffer_size = 255
                required_string_buffer_size = ctypes.c_size_t(0)
                string_buffers = [
                    ctypes.create_string_buffer(reserved_string_buffer_size) for i in range_(self.__num_inner_eval)
                ]
                ptr_string_buffers = (ctypes.c_char_p * self.__num_inner_eval)(*map(ctypes.addressof, string_buffers))
                _safe_call(_LIB.LGBM_BoosterGetEvalNames(
                    self.handle,
                    self.__num_inner_eval,
                    ctypes.byref(tmp_out_len),
                    reserved_string_buffer_size,
                    ctypes.byref(required_string_buffer_size),
>                   ptr_string_buffers))
E               WindowsError: exception: access violation writing 0x00007FF9085F04DE

C:\Miniconda\envs\test-env\lib\site-packages\lightgbm\basic.py:3182: WindowsError

```

**Pull request merged**